server/src/test/java/org/elasticsearch/transport/LeakTrackerTests.java

```diff
@@ -0,0 +1,206 @@
+        }
+
+        @Override
+        public void nullifyReference() {
+            object = null;
+        }
+
+        @Override
+        public void closeLeak() {
+            leak.close();
+        }
+
+        @Override
+        public String toString() {
+            return "LeakTracker.INSTANCE.track(Object)";
+        }
+    }
+
+    private static class ReferenceCountedTrackedObjectTestCase implements TrackedObjectTestCase {
+
+        private RefCounted refCounted;
+
+        @Override
+        public void makeAssumptions() {
+            assumeTrue("ReferenceCounted wrapper is a no-op when assertions are disabled", Assertions.ENABLED);
+        }
+
+        @Override
+        public void createAndTrack(ReachabilityChecker reachabilityChecker) {
+            RefCounted refCounted = reachabilityChecker.register(createRefCounted());
+            this.refCounted = LeakTracker.wrap(refCounted);
+            this.refCounted.incRef();
+            this.refCounted.tryIncRef();
+        }
+
+        @Override
+        public void nullifyReference() {
+            refCounted = null;
+        }
+
+        @Override
+        public void closeLeak() {
+            refCounted.decRef();    // tryIncRef
+            refCounted.decRef();    // incRef
+            refCounted.decRef();    // implicit
+        }
+
+        @Override
+        public String toString() {
+            return "LeakTracker.wrap(RefCounted)";
+        }
+
+        private static RefCounted createRefCounted() {
+            int number = Randomness.get().nextInt();
+            return new AbstractRefCounted() {
+
+                @Override
+                protected void closeInternal() {
+                    // Do nothing
+                    logger.info("Prevents this returning a non-collectible constant {}", number);
+                }
+            };
+        }
+    }
+
+    private static class ReleasableTrackedObjectTestCase implements TrackedObjectTestCase {
+
+        private Releasable releasable;
+
+        @Override
+        public void makeAssumptions() {
+            assumeTrue("Releasable wrapper is a no-op when assertions are disabled", Assertions.ENABLED);
+        }
+
+        @Override
+        public void createAndTrack(ReachabilityChecker reachabilityChecker) {
+            Releasable releasable = reachabilityChecker.register(createReleasable());
+            this.releasable = LeakTracker.wrap(releasable);
+        }
+
+        @Override
+        public void nullifyReference() {
+            releasable = null;
+        }
+
+        @Override
+        public void closeLeak() {
+            releasable.close();
+        }
+
+        @Override
+        public String toString() {
+            return "LeakTracker.wrap(Releasable)";
+        }
+
+        private static Releasable createReleasable() {
+            int number = Randomness.get().nextInt();
+            return () -> logger.info("Prevents this returning a non-collectible constant {}", number);
```

Comment:
I originally returned `() -> {}` but for some reason (I think the compiler turned it into a constant) it would never get garbage collected that way. Adding the random parameter is one way of ensuring that doesn't happen.

Reply:
Yeah `() -> {}` is a static constant. I think `Releasables.assertOnce(()->{})` would work tho.